docs: rewrite Configure Agents index and document external agent harnesses (#2311)
NVIDIA-NeMo/Gym: fern/versions/latest/pages/agent-server/index.mdx

---
title: "Configure Agents"

> description: "The Agent server holds an environment's agent harness. Use a built-in agent, an agent that wraps an external harness, or your own."

Reviewer: ```suggestion
he Agent server holds the agent harness. Use a built-in agent, an agent that wraps an external harness, or your own.
```

position: 1
---

> An environment is made of four parts: a dataset, an agent harness, a verifier, and per-task state. The Agent server is where the harness lives.

Reviewer: ```suggestion
An environment integrates four independent parts: a dataset, an agent harness, a verifier, and per-task state. The Agent server is where the harness lives.
```

A harness decides how a model interacts with the world: how many turns it gets, what its system prompt says, which tools it can reach, how tool results are fed back, when it stops, and what scaffolding (planning, compaction, sub-agents) sits between the model and the task. The Agent server does not run an LLM itself. It is orchestration code that delegates text generation to the Model server and tool execution, state, and scoring to the Resources server.

## Why the Harness Is Its Own Server

> Because the harness is a separate server rather than part of the dataset or the verifier, it can be changed independently of the rest of the environment:

Reviewer: ```suggestion
Because if the harness is a separate server rather than part of the dataset or the verifier, it can be changed independently of the rest of the environment:
```

- Run several harnesses against one environment, with the dataset and verifier held fixed.

> - Run one harness across many environments. The Agent server references a Resources server by name, so switching benchmarks is a config change.

Reviewer: ```suggestion
- Run one harness in many environments. The Agent server is linked to the Resources server by name, so switching benchmarks is a config change.
```

        return conversation
```

> An agent that wraps an external harness replaces the middle phase with the harness's own loop. The outer shape is the same: NeMo Gym seeds the session, verifies the finished trajectory, and reports the reward.

Reviewer: What do you mean by middle phase here?

```suggestion
An agent server that is composed by an external harness basically is replacing the inner interaction loop with the policy model. The episode orchestration is the same: NeMo Gym's seeds the session, verifies the finished trajectory, and reports the reward.
```

> ## Choosing an Agent

Reviewer: ```suggestion
## Choosing an Agent Server to implement your agent
```

> There are three options. Use a built-in NeMo Gym agent — `simple_agent` handles multi-step tool calling against any Resources server. Use one of the agents that wraps an external harness, such as a coding CLI, an agent framework, or a benchmark's own scaffold; these need configuration but no code. Or keep your agent in your own repo and expose an endpoint that follows the OpenAI Responses contract, and let `remote_agent` drive the loop against it.

Reviewer: ```suggestion
There are three options. Use a built-in NeMo Gym agent server — `simple_agent` handles multi-step tool calling against any Resources server. Use one of the agent servers that wraps an external harness, such as a coding CLI, an agent framework, or a benchmark's own scaffold; these need configuration but no code. Or keep your the agent server in your own repo and expose an endpoint that follows the OpenAI Responses contract, and let `remote_agent` server drive the loop against it.
```

<Cards>

<Card title="Integrate Existing Agents" href="/agent-server/integrate-existing-agents">
Built-in and external agents, how each external harness is wrapped, and what a wrapper needs to support training.
</Card>

<Card title="Drive a Remote Agent" href="/agent-server/remote-agent">
The `remote_agent` contract, an end-to-end quickstart, configuration fields, and failure handling.
</Card>

<Card title="Agent Skills" href="/agent-server/agent-skills">
Set [Agent Skills](https://agentskills.io/specification) per run, separately from the dataset.
</Card>

<Card title="Agent Server Fields" href="/reference/configuration#agent-server-fields">

> Configuration syntax for wiring an agent to its Resources server and Model server.

Reviewer: ```suggestion
Configuration syntax for wiring an agent server to a compatible Resources server and Model server.
```
